src/modules/esl-carousel/core/esl-carousel.ts

```diff
@@ -0,0 +1,302 @@
+    });
+  }
+
+  /** @returns currently active renderer */
+  @memoize()
+  public get renderer(): ESLCarouselRenderer {
+    return ESLCarouselRenderer.registry.create(this, this.configCurrent);
+  }
+
+  @ready
+  protected override connectedCallback(): void {
+    super.connectedCallback();
+    this.update();
+    this.updateA11y();
+  }
+
+  protected override attributeChangedCallback(attrName: string, oldVal: string, newVal: string): void {
+    if (!this.connected) return;
+    memoize.clear(this, `${attrName}Rule`);
+    this.update();
+  }
+
+  protected override disconnectedCallback(): void {
+    super.disconnectedCallback();
+    memoize.clear(this, ['$container', '$slides', '$slidesArea']);
+  }
+
+  /** Updates the config and the state that is associated with */
+  @decorate(microtask)
+  public update(): void {
+    const config = this.configCurrent;
+    const oldConfig = this.config;
+    const $oldSlides = this.$slides;
+
+    memoize.clear(this, '$slides');
+    const added = this.$slides.filter((slide) => !$oldSlides.includes(slide));
+    const removed = $oldSlides.filter((slide) => !this.$slides.includes(slide));
+
+    if (!added.length && !removed.length && this.renderer.equal(config)) return;
+
+    this.renderer.unbind();
+    memoize.clear(this, 'renderer');
+    this.renderer && this.renderer.bind();
+    this.updateContainer();
+    this.dispatchEvent(ESLCarouselChangeEvent.create({added, removed, config, oldConfig}));
+  }
+
+  public updateContainer(): void {
+    if (!this.$container) return;
+    this.$container.toggleAttribute('empty', this.size === 0);
+    this.$container.toggleAttribute('single', this.size === 1);
+    this.$container.toggleAttribute('incomplete', this.size <= this.renderer.count);
+  }
+
+  /** Appends slide instance to the current carousel */
+  public addSlide(slide: HTMLElement | ESLCarouselSlide): void {
+    if (!slide) return;
+    if (!(slide instanceof ESLCarouselSlide)) return this.addSlide(ESLCarouselSlide.create(slide));
+    if (slide.parentNode !== this.$slidesArea) {
+      slide.remove();
+      this.$slidesArea?.appendChild(slide);
+    }
+    this.update();
+  }
+
+  /** Remove slide instance from the current carousel */
+  public removeSlide(slide: HTMLElement | ESLCarouselSlide): void {
+    if (!slide) return;
+    if (!(slide instanceof ESLCarouselSlide)) return this.removeSlide(slide.closest(ESLCarouselSlide.is)!);
+    if (slide.parentNode === this.$slidesArea) this.$slidesArea?.removeChild(slide);
+    this.update();
+  }
+
+  protected updateA11y(): void {
+    const $container = this.$container || this;
+    if (!$container.hasAttribute('role')) {
+      $container.setAttribute('role', 'region');
+      $container.setAttribute('aria-roledescription', 'Carousel');
+    }
+    if (!this.$slidesArea.hasAttribute('role')) {
+      this.$slidesArea.setAttribute('role', 'list');
+      // this.$slidesArea.setAttribute('aria-live', 'polite');
```

Comment:
comment

Reply:
Let's leave it for now. We need to revisit that idea from a11ty perspective, so it will be nice to have it here as a reminder for now.